pyogrio/raw.py

```diff
@@ -349,10 +359,11 @@ def open_arrow(
     sql_dialect=None,
     return_fids=False,
     batch_size=65_536,
+    return_pyarrow=True,
```

Comment:
It doesn't return a capsule _exactly_, but a small wrapper object with the `__arrow_c_stream__` method that will return the capsule. 

Side question: would returning the capsule directly be more useful for you?

Reply:
(although it would be nice to have keyword name that is "positive" about what you want, instead of indicating of _not_ wanting pyarrow .. I just don't have good ideas for that at the moment)

Reply:
> (although it would be nice to have keyword name that is "positive" about what you want, instead of indicating of _not_ wanting pyarrow .. I just don't have good ideas for that at the moment)

I think this is more along the lines of what I was trying to ask. 

I think having a wrapper object is preferable than raw capsules

Reply:
We could maybe also consider removing the option entirely, and always returns the capsule wrapper object. This API is still quite new and more an advanced API, so we might be OK with still changing it.

Taking the example from the docstring:

```python
with open_arrow(path) as source:
    meta, reader = source
    for table in reader:
        ....
```

If we would just always return the wrapper object as implemented in the PR right now, that means that people using this right now and expecting a pyarrow RecordBachReader (or an iterable of pyarrow Tables), would need to change their code to:

```
with open_arrow(path) as source:
    meta, reader = source
    reader = pa.RecordBatchReader.from_stream(reader)
    for table in reader:
        ....
```

One problem with this however is that this requires pyarrow >= 15.0, so that's not ideal.

We could also add the basic methods of a RecordBatchreader to the wrapper object (`__enter__`, `__iter__`, `read_next_batch`, `read_all`, `schema`, etc, just calling the equivalent method of the underlying pyarrow.RecordBatchReader, which we can instantiate lazily only when one of those methods is used, i.e. which still allows you to convert it with `__arrow_c_stream__` without requiring pyarrow)

That keeps compatibility for people using the Python APIs of a RecordBatchReader (like in the code snippet above, to iterate through the stream). However, that unfortunately still doesn't work when passing this reader object to some other functionality of pyarrow that can accept a reader but passes that through to C++ code (eg writing a dataset with a RecordBatchReader as input)

Reply:
I don't know what the best option is here. I'm ok with an option to return either, and maybe sometime in the future (when pyarrow >=15 is an ok dependency constraint) we could make a breaking change?

For my own uses, [calling this from Rust](https://github.com/geoarrow/geoarrow-rs/blob/0e51bfc3f386e888a8b34df6c6c0c162d08f3bda/python/core/src/interop/pyogrio/from_pyogrio.rs#L146-L156), I just expect the input to have an `__arrow_c_stream__` dunder. Most people will have pyarrow installed, and so that's an ok default.

Reply:
We could also just check if pyarrow is available, if so return a RecordBatchReader (which will also have the `__arrow_c_stream__` method for recent pyarrow), and if not fallback on returning the wrapper object. 

That keeps backwards compatibility perfectly, while allowing to use it without pyarrow (at the expense of a slightly confusing API / different behaviour just depending on whether another package is installed or not)

Although that might also be an extra indirection in case you don't need the pyarrow RecordBatchReader. Because what will be returned from `__arrow_c_stream__` on the returned reader will be a ArrowArrayStream backed by Arrow wrapping the GDAL stream.

Reply:
I don't think the extra indirection would be a problem.

I tend to think that having multiple possible return types is much more confusion than it's worth for most users. I'm not really a fan of having it depend only on if pyarrow is installed.

If we're going to have one function, having a `return_pyarrow=True` I think is the best option. It's probably too much duplication here, but otherwise I would've considered an `open_arrow_pycapsule_interface()` or similar function

Reply:
> We could also just check if pyarrow is available, if so return a RecordBatchReader (which will also have the `__arrow_c_stream__` method for recent pyarrow), and if not fallback on returning the wrapper object.

Silently returning a different type depending on installed Python packages is generally error-prone for users. I would recommend avoiding this.

Reply:
Good points, agreed that it's better to not do any "smart" return value depending on what is installed.

I updated the PR (and added some docs) for the current state of having a `return_pyarrow` keyword with a default of True.

I am personally also still fine with changing the default to False (so you have to ask explicitly for a pyarrow object (to avoid the one line to do it yourself), which makes the keyword also about what you want, not what you want to avoid)

Reply:
The keyword mentioned in the changelog is `use_pyarrow` (which I think I prefer) but the implementation is `return_pyarrow`